File: sfrog/crawler/parser.py
"""HTML parsing utilities based on :mod:`selectolax`."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, List, Optional, Tuple

from selectolax.parser import HTMLParser

from sfrog.utils.url_tools import is_same_domain, normalize_url


Heading = Tuple[str, str]


@dataclass(slots=True)
class ParsedPage:
    url: str
    title: Optional[str]
    meta_description: Optional[str]
    canonical: Optional[str]
    headings: List[Heading]
    internal_links: List[str]
    external_links: List[str]
    images: List[str]


def parse_html(url: str, base_url: str, html: str) -> ParsedPage:
    """Parse HTML content extracting SEO signals."""

    parser = HTMLParser(html)
    title = _first_text(parser.css("title"))
    meta_description = None
    for node in parser.css("meta"):
        if node.attributes.get("name", "").lower() == "description":
            meta_description = node.attributes.get("content")
            break
    canonical = None
    for node in parser.css("link"):
        if node.attributes.get("rel", "").lower() == "canonical":
            canonical = normalize_url(url, node.attributes.get("href", ""))
            break

    headings: List[Heading] = []
    for node in parser.css("h1, h2, h3, h4, h5, h6"):
        text = node.text(strip=True)
        if text:
            headings.append((node.tag.lower(), text))

    internal_links: List[str] = []
    external_links: List[str] = []
    for node in parser.css("a"):
        href = node.attributes.get("href")
        if not href:
            continue
        normalized = normalize_url(url, href)
        if is_same_domain(normalized, base_url):
            internal_links.append(normalized)
        else:
            external_links.append(normalized)

    images = [node.attributes.get("alt", "") for node in parser.css("img") if node.attributes.get("alt")]

    return ParsedPage(
        url=url,
        title=title,
        meta_description=meta_description,
        canonical=canonical,
        headings=headings,
        internal_links=internal_links,
        external_links=external_links,
        images=images,
    )


def _first_text(nodes: Iterable[HTMLParser]) -> Optional[str]:
    for node in nodes:
        text = node.text(strip=True)
        if text:
            return text
    return None


__all__ = ["ParsedPage", "Heading", "parse_html"]

File: internal/crawler/parser.go
package crawler

import (
	"net/url"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/dillonlara115/baracuda/internal/utils"
	"github.com/dillonlara115/baracuda/pkg/models"
)

// Parser extracts SEO data from HTML content
type Parser struct {
	baseURL string
	domain  string
}

// NewParser creates a new Parser instance
func NewParser(baseURL string) (*Parser, error) {
	domain, err := utils.ExtractDomain(baseURL)
	if err != nil {
		return nil, err
	}

	return &Parser{
		baseURL: baseURL,
		domain:  domain,
	}, nil
}

// Parse extracts SEO data from HTML content
func (p *Parser) Parse(htmlContent []byte) (*models.PageResult, error) {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(string(htmlContent)))
	if err != nil {
		return nil, err
	}

	result := &models.PageResult{
		URL:           p.baseURL,
		H1:            make([]string, 0),
		H2:            make([]string, 0),
		H3:            make([]string, 0),
		H4:            make([]string, 0),
		H5:            make([]string, 0),
		H6:            make([]string, 0),
		InternalLinks: make([]string, 0),
		ExternalLinks: make([]string, 0),
		Images:        make([]models.Image, 0),
	}

	// Extract title
	result.Title = strings.TrimSpace(doc.Find("title").First().Text())

	// Extract meta description
	doc.Find("meta[name='description']").Each(func(i int, s *goquery.Selection) {
		if content, exists := s.Attr("content"); exists {
			result.MetaDesc = strings.TrimSpace(content)
		}
	})

	// Extract canonical link
	doc.Find("link[rel='canonical']").Each(func(i int, s *goquery.Selection) {
		if href, exists := s.Attr("href"); exists {
			result.Canonical = strings.TrimSpace(href)
		}
	})

	// Extract headings
	doc.Find("h1").Each(func(i int, s *goquery.Selection) {
		text := strings.TrimSpace(s.Text())
		if text != "" {
			result.H1 = append(result.H1, text)
		}
	})
	doc.Find("h2").Each(func(i int, s *goquery.Selection) {
		text := strings.TrimSpace(s.Text())
		if text != "" {
			result.H2 = append(result.H2, text)
		}
	})
	doc.Find("h3").Each(func(i int, s *goquery.Selection) {
		text := strings.TrimSpace(s.Text())
		if text != "" {
			result.H3 = append(result.H3, text)
		}
	})
	doc.Find("h4").Each(func(i int, s *goquery.Selection) {
		text := strings.TrimSpace(s.Text())
		if text != "" {
			result.H4 = append(result.H4, text)
		}
	})
	doc.Find("h5").Each(func(i int, s *goquery.Selection) {
		text := strings.TrimSpace(s.Text())
		if text != "" {
			result.H5 = append(result.H5, text)
		}
	})
	doc.Find("h6").Each(func(i int, s *goquery.Selection) {
		text := strings.TrimSpace(s.Text())
		if text != "" {
			result.H6 = append(result.H6, text)
		}
	})

	// Extract links
	doc.Find("a[href]").Each(func(i int, s *goquery.Selection) {
		href, exists := s.Attr("href")
		if !exists {
			return
		}

		// Resolve relative URLs
		resolvedURL, err := utils.ResolveURL(p.baseURL, href)
		if err != nil {
			return
		}

		// Normalize URL
		normalizedURL, err := utils.NormalizeURL(resolvedURL)
		if err != nil {
			return
		}

		// Skip fragments, javascript:, mailto:, etc.
		u, err := url.Parse(normalizedURL)
		if err != nil {
			return
		}

		if u.Scheme != "http" && u.Scheme != "https" {
			return
		}

		// Categorize as internal or external
		if utils.IsSameDomain(normalizedURL, p.baseURL) {
			// Avoid duplicates
			for _, existing := range result.InternalLinks {
				if existing == normalizedURL {
					return
				}
			}
			result.InternalLinks = append(result.InternalLinks, normalizedURL)
		} else {
			// Avoid duplicates
			for _, existing := range result.ExternalLinks {
				if existing == normalizedURL {
					return
				}
			}
			result.ExternalLinks = append(result.ExternalLinks, normalizedURL)
		}
	})

	// Extract images
	doc.Find("img").Each(func(i int, s *goquery.Selection) {
		src, exists := s.Attr("src")
		if !exists {
			return
		}

		// Resolve relative URLs
		resolvedURL, err := utils.ResolveURL(p.baseURL, src)
		if err != nil {
			return
		}

		// Normalize URL
		normalizedURL, err := utils.NormalizeURL(resolvedURL)
		if err != nil {
			return
		}

		// Skip data URIs and non-HTTP schemes
		u, err := url.Parse(normalizedURL)
		if err != nil {
			return
		}

		if u.Scheme != "http" && u.Scheme != "https" {
			return
		}

		// Get alt text
		alt := s.AttrOr("alt", "")

		// Avoid duplicates
		for _, existing := range result.Images {
			if existing.URL == normalizedURL {
				return
			}
		}

		result.Images = append(result.Images, models.Image{
			URL: normalizedURL,
			Alt: alt,
		})
	})

	return result, nil
}

// ExtractLinks extracts all links from HTML content and returns them as a slice
func (p *Parser) ExtractLinks(htmlContent []byte) ([]string, error) {
	result, err := p.Parse(htmlContent)
	if err != nil {
		return nil, err
	}

	links := make([]string, 0, len(result.InternalLinks)+len(result.ExternalLinks))
	links = append(links, result.InternalLinks...)
	links = append(links, result.ExternalLinks...)

	return links, nil
}

